styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -127,38 +135,110 @@
   private final Stats stats;
   private final int pollPodsIntervalSeconds;
   private final ServiceAccountKeyManager serviceAccountKeyManager;
+  private final Supplier<String> secretEpochProvider;
 
   private Watch watch;
 
   private final Object secretMutationLock = new Object() {};
 
   KubernetesDockerRunner(NamespacedKubernetesClient client, StateManager stateManager, Stats stats,
-      ServiceAccountKeyManager serviceAccountKeyManager, int pollPodsIntervalSeconds) {
+      ServiceAccountKeyManager serviceAccountKeyManager, int pollPodsIntervalSeconds,
+      Supplier<String> secretEpochProvider) {
     this.stateManager = Objects.requireNonNull(stateManager);
     this.client = Objects.requireNonNull(client).inNamespace(NAMESPACE);
     this.stats = Objects.requireNonNull(stats);
     this.pollPodsIntervalSeconds = pollPodsIntervalSeconds;
-    this.serviceAccountKeyManager = serviceAccountKeyManager;
+    this.serviceAccountKeyManager = Objects.requireNonNull(serviceAccountKeyManager);
+    this.secretEpochProvider = Objects.requireNonNull(secretEpochProvider);
   }
 
   KubernetesDockerRunner(NamespacedKubernetesClient client, StateManager stateManager, Stats stats,
       ServiceAccountKeyManager serviceAccountKeyManager) {
-    this(client, stateManager, stats, serviceAccountKeyManager, DEFAULT_POLL_PODS_INTERVAL_SECONDS);
+    this(client, stateManager, stats, serviceAccountKeyManager, DEFAULT_POLL_PODS_INTERVAL_SECONDS,
+        DEFAULT_SECRET_EPOCH_PROVIDER);
   }
 
   @Override
   public String start(WorkflowInstance workflowInstance, RunSpec runSpec) throws IOException {
-    ensureSecrets(workflowInstance, runSpec);
+    final String secretEpoch = secretEpochProvider.get();
+    synchronized (secretMutationLock) {
+      ensureSecrets(workflowInstance, runSpec, secretEpoch);
+      try {
+        final Pod pod = client.pods().create(createPod(workflowInstance, runSpec, secretEpoch));
+        return pod.getMetadata().getName();
+      } catch (KubernetesClientException kce) {
+        throw new IOException("Failed to create Kubernetes pod", kce);
+      }
+    }
+  }
+
+  @Override
+  public void cleanup() throws IOException {
+    synchronized (secretMutationLock) {
+
+      // Enumerate all secrets currently used by pods
+      final PodList pods = client.pods().list();
+      final Set<String> activeSecrets = pods.getItems().stream()
+          .flatMap(pod -> pod.getSpec().getVolumes().stream())
+          .map(volume -> volume.getSecret().getSecretName())
+          .collect(Collectors.toSet());
+
+      // Enumerate all service account secrets that are not in active use
+      final SecretList secrets = client.secrets().list();
+      final List<Secret> inactiveServiceAccountSecrets = secrets.getItems().stream()
+          .filter(secret -> secret.getMetadata().getName().startsWith(STYX_WORKFLOW_SA_SECRET_NAME))
```

Comment:
Right, I guess that will make the key rotation less aggressive and cause less key churn, which seems like a good idea given that the IAM write request quota is not very high out of the box.